text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,248 @@
+to the intermediary.
+
+#### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new messages.
+
+The consumer receives, processes, and settles a message.
+
+* "Receiving" is the process of obtaining a message from the intermediary.
+* "Processing" is the process of acting on the information a message contains.
+* "Settling" is the process where intermediary and consumer agree on the state
+  of the transfer.
+
+#### Intermediary
+
+An "intermediary" receives a message for the purpose of forwarding it to the
+next receiver, which might be another intermediary or a consumer.
+
+### Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```
+                               CONSUMER
+
+                   . . . . . . Settle
+PRODUCER           .              ^
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extraced from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+> A producer SHOULD attach a creation context to each message. The creation context
+> SHOULD be attached in a way so that it is not changed by intermediaries.
+
+### Trace structure, names, and attributes
+
```

Comment:
For producer instrumentation, @blumamir identified several different possiblities.

1. Having a dedicated span and context for each single message, also for batched messages. There could be a single "Publish" span for a batch, but there needs to be a single "Message" span for each single message in the span. The context of the "Message" span will be passed on and used to link consumer traces.
2. Having one ("Publish") span per batch, and storing message specific information on this single span.
   a. Message specific information could be stored in events that are added to the span.
   b. Message specific information could be stored in an attribute (e. g. an attribute that consists of a list of message ids).
3. Having one "Publish" span for a batch, but having a dedicated span-less context for each message. This span-less context will be passed on with each message and will be used to link consumer traces.